<a href="https://colab.research.google.com/github/sajid-munawar/100_plus_python_challenging_programs/blob/master/100%2B_Python_challenges.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

100+ Python challenging programming exercises

Level description


Level 1	Beginner:

Means someone who has just gone through an introductory Python course. He can solve some problems with 1 or 2 Python classes or functions. Normally, the answers could directly be found in the textbooks.

Level 2 Intermediate:

Means someone who has just learned Python, but already has a relatively strong programming background from before. He should be able to solve problems which may involve 3 or 3 Python classes or functions. The answers cannot be directly be found in the textbooks.

Level 3	Advanced:

He should use Python to solve more complex problem using more rich libraries functions and data structures and algorithms. He is supposed to solve the problem using several Python standard packages and advanced techniques.



2.	Problem template

#----------------------------------------#
Question

Hints

Solution

3.	Questions

#----------------------------------------#

In [ ]:
%4096

1

In [ ]:
def calculate_storage(filesize):
    block_size = 4096
    # Use floor division to calculate how many blocks are fully occupied
    full_blocks = filesize//block_size
    # Use the modulo operator to check whether there's any remainder
    partial_block_remainder = 
    # Depending on whether there's a remainder or not, return
    # the total number of bytes required to allocate enough blocks
    # to store your data.
    if partial_block_remainder >0:
        return block_size*2
    return block_size

print(calculate_storage(1))    # Should be 4096
print(calculate_storage(4096)) # Should be 4096
print(calculate_storage(4097)) # Should be 8192
print(calculate_storage(6000)) # Should be 8192

8192
4096
8192
8192


In [ ]:
10//10

1

In [ ]:
def calculate_storage(filesize):
    block_size = 4096
    # Use floor division to calculate how many blocks are fully occupied
    full_blocks = filesize//block_size
    # Use the modulo operator to check whether there's any remainder
    partial_block_remainder =% 
    # Depending on whether there's a remainder or not, return
    # the total number of bytes required to allocate enough blocks
    # to store your data.
    if partial_block_remainder >0:
        return block_size*2
    return block_size

print(calculate_storage(1))    # Should be 4096
print(calculate_storage(4096)) # Should be 4096
print(calculate_storage(4097)) # Should be 8192
print(calculate_storage(6000)) # Should be 8192

10
20
20
20


In [ ]:
file_size=int(input('file size'))
block_size=4096
full_blocks=file_size%block_size
full_blocks
# partial_block_remainder=
# if partial_block_remainder >0:
  # return 4096

file size4444


348

In [ ]:
"big" > "small"

False

In [ ]:
def sum(x, y):
		return(x+y)
print(sum(sum(1,2), sum(3,4)))

10


In [ ]:
((10 >= 5*2) and (10 <= 5*2))

True


Question 1
Level 1

Question:
Write a program which will find all such numbers which are divisible by 7 but are not a multiple of 5,
between 2000 and 3200 (both included).
The numbers obtained should be printed in a comma-separated sequence on a single line.

In [ ]:
for i in range(2000,3200+1):
  if i%7==0 and i%5!=0:
    print(i,end=',')

2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199,

In [ ]:
alist=[]
for i in range(2000,3200+1):
  if i%7==0 and i%5!=0:
    alist.append(i)
  # print(i,end=' ')
print(alist)

[2002, 2009, 2016, 2023, 2037, 2044, 2051, 2058, 2072, 2079, 2086, 2093, 2107, 2114, 2121, 2128, 2142, 2149, 2156, 2163, 2177, 2184, 2191, 2198, 2212, 2219, 2226, 2233, 2247, 2254, 2261, 2268, 2282, 2289, 2296, 2303, 2317, 2324, 2331, 2338, 2352, 2359, 2366, 2373, 2387, 2394, 2401, 2408, 2422, 2429, 2436, 2443, 2457, 2464, 2471, 2478, 2492, 2499, 2506, 2513, 2527, 2534, 2541, 2548, 2562, 2569, 2576, 2583, 2597, 2604, 2611, 2618, 2632, 2639, 2646, 2653, 2667, 2674, 2681, 2688, 2702, 2709, 2716, 2723, 2737, 2744, 2751, 2758, 2772, 2779, 2786, 2793, 2807, 2814, 2821, 2828, 2842, 2849, 2856, 2863, 2877, 2884, 2891, 2898, 2912, 2919, 2926, 2933, 2947, 2954, 2961, 2968, 2982, 2989, 2996, 3003, 3017, 3024, 3031, 3038, 3052, 3059, 3066, 3073, 3087, 3094, 3101, 3108, 3122, 3129, 3136, 3143, 3157, 3164, 3171, 3178, 3192, 3199]


In [ ]:
alist1=[str(y) for y in range(2000,3201) if y%7==0 if y%5!=0]
print(",".join((alist1)))
# print(alist1)

2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199


In [ ]:


# Hints:
# Consider use range(#begin, #end) method

# Solution:
l=[]
for i in range(2000, 3201):
    if (i%7==0) and (i%5!=0):
        l.append(str(i))

print(','.join(l))
#----------------------------------------#

2002,,2009,,2016,,2023,,2037,,2044,,2051,,2058,,2072,,2079,,2086,,2093,,2107,,2114,,2121,,2128,,2142,,2149,,2156,,2163,,2177,,2184,,2191,,2198,,2212,,2219,,2226,,2233,,2247,,2254,,2261,,2268,,2282,,2289,,2296,,2303,,2317,,2324,,2331,,2338,,2352,,2359,,2366,,2373,,2387,,2394,,2401,,2408,,2422,,2429,,2436,,2443,,2457,,2464,,2471,,2478,,2492,,2499,,2506,,2513,,2527,,2534,,2541,,2548,,2562,,2569,,2576,,2583,,2597,,2604,,2611,,2618,,2632,,2639,,2646,,2653,,2667,,2674,,2681,,2688,,2702,,2709,,2716,,2723,,2737,,2744,,2751,,2758,,2772,,2779,,2786,,2793,,2807,,2814,,2821,,2828,,2842,,2849,,2856,,2863,,2877,,2884,,2891,,2898,,2912,,2919,,2926,,2933,,2947,,2954,,2961,,2968,,2982,,2989,,2996,,3003,,3017,,3024,,3031,,3038,,3052,,3059,,3066,,3073,,3087,,3094,,3101,,3108,,3122,,3129,,3136,,3143,,3157,,3164,,3171,,3178,,3192,,3199


#----------------------------------------#
Question 2
Level 1

Question:

Write a program which can compute the factorial of a given numbers.
The results should be printed in a comma-separated sequence on a single line.
Suppose the following input is supplied to the program:

8

Then, the output should be:

40320

In [ ]:
factorial=1
for n in range(1,9):
  factorial=factorial*n
print(factorial)

40320


In [ ]:
num=int(input('please enter num'))
a=1
for i in range(1,num+1):
  a*=i
  print(a,end=',')

please enter num8
1,2,6,24,120,720,5040,40320,

In [ ]:
alist=[]
num=int(input('please enter num'))
a=1
for i in range(1,num+1):
  a*=i
  alist.append(str(a))
print(','.join(alist))

please enter num8
1,2,6,24,120,720,5040,40320


In [ ]:
(lambda x:x*x-1)(7)

48

In [ ]:
1*2*3*4*5*6*7*8

40320

In [ ]:
# Hints:
# In case of input data being supplied to the question, it should be assumed to be a console input.

# Solution:
def fact(x):
    if x == 0:
        return 1
    return x * fact(x - 1)

x=int(input())
print(fact(x))
#----------------------------------------#

8
40320


In [ ]:
def factorial(num):
  y=1
  while num>1:
    y*=num
    num-=1
  print(y)

In [ ]:
def abc(x):
  if x==0:
    return 1
  # abc(x-4)
  return x*fact(x-1)

In [ ]:
# fact(968) max limit

#----------------------------------------#
Question 3
Level 1

Question:

With a given integral number n, write a program to generate a dictionary that contains (i, i*i) such that is an integral number between 1 and n (both included). and then the program should print the dictionary.
Suppose the following input is supplied to the program:

8

Then, the output should be:

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}

In [ ]:
def dic(x):
  dictionary={}
  for i in range(1,x+1):
    dictionary[i]=i*i
  print(dictionary)

In [ ]:
dic(8)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}


In [ ]:
# Hints:
# In case of input data being supplied to the question, it should be assumed to be a console input.
# Consider use dict()

# Solution:
n=int(input())
d=dict()
for i in range(1,n+1):
    d[i]=i*i

print(d)
#----------------------------------------#

8
{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}


In [ ]:
#----------------------------------------#
Question 4
Level 1

Question:

Write a program which accepts a sequence of comma-separated numbers from console and generate a list and a tuple which contains every number.
Suppose the following input is supplied to the program:

34,67,55,33,12,98

Then, the output should be:

['34', '67', '55', '33', '12', '98']
('34', '67', '55', '33', '12', '98')


In [ ]:
values=input()
alist=values.split(',')
tup=tuple(alist)
print(alist)
print(tup)

22,33,44,55
['22', '33', '44', '55']
('22', '33', '44', '55')


In [ ]:
# Hints:
# In case of input data being supplied to the question, it should be assumed to be a console input.
# tuple() method can convert list to tuple

# Solution:
values=input()
l=values.split(",")
t=tuple(l)
print(l)
print(t)
#----------------------------------------#


33,44,55
['33', '44', '55']
('33', '44', '55')


#----------------------------------------#
Question 5
Level 1

Question:

Define a class which has at least two methods:

getString: 

to get a string from console input


printString: 

to print the string in upper case.

Also please include simple test function to test the class methods.

In [ ]:
class Car():
  def __init__(self):
    self=self

  def getString(self):
    self.string=input('Please enter the string ')
  def printString(self):
    print(self.string.upper())

In [ ]:
car1=Car()

In [ ]:
car1.getString()

Please enter the string my name is sajid


In [ ]:
car1.printString()

MY NAME IS SAJID


In [ ]:
# Hints:
# Use __init__ method to construct some parameters

# Solutionmy:
class InputOutString():
    def __init__(self):
        self.s = ""

    def getString(self):
        self.s =input()

    def printString(self):
        print(self.s.upper())

strObj = InputOutString()
strObj.getString()
strObj.printString()
#----------------------------------------#

my name is Sajid
MY NAME IS SAJID


#----------------------------------------#
'''Question 6
Level 2

Question:
Write a program that calculates and prints the value according to the given formula:
Q = Square root of [(2 * C * D)/H]
Following are the fixed values of C and H:
C is 50. H is 30.
D is the variable whose values should be input to your program in a comma-separated sequence.
Example
Let us assume the following comma separated input sequence is given to the program:
100,150,180
The output of the program should be:
18,22,24
'''

In [ ]:
def formula():
  valuess=input()
  final=[]
  alist=valuess.split(',')
  for i in range(len(alist)):
    final.append(str(int(sqrt(2*50*int((alist[i]))//30))))
  print(','.join(final))
  # return alist

In [ ]:
'''
Hints:
If the output received is in decimal form, it should be rounded off to its nearest value (for example, if the output received is 26.0, it should be printed as 26)
In case of input data being supplied to the question, it should be assumed to be a console input. 

Solution:
#!/usr/bin/env python'''
import math
c=50
h=30
value = []
items=[x for x in input().split(',')]
for d in items:
    value.append(str(int(round(math.sqrt(2*c*float(d)/h)))))

print(','.join(value))
#----------------------------------------#

100,150,180
18,22,24


'''#----------------------------------------#
Question 7

Level 2

Question:

Write a program which takes 2 digits, X,Y as input and generates a 2-dimensional array.
 The element value in the i-th row and j-th column of the array should be i*j.
Note: i=0,1.., X-1; j=0,1,¡­Y-1.
Example

Suppose the following inputs are given to the program:
3,5

Then, the output of the program should be:

[[0, 0, 0, 0, 0],

 [0, 1, 2, 3, 4],

 [0, 2, 4, 6, 8]] 

'''

In [ ]:
import numpy as np
digits=[x for x in input().split(',')]
arr=np.zeros((int(digits[0]),int(digits[1])))
# print(arr)
for row in range((arr.shape[0])):
  for column in range(arr.shape[1]):
    arr[row][column]=row*column
print(arr)

3,5
[[0. 0. 0. 0. 0.]
 [0. 1. 2. 3. 4.]
 [0. 2. 4. 6. 8.]]


In [ ]:
'''Hints:
Note: In case of input data being supplied to the question, it should be assumed to be a console input in a comma-separated form.

Solution:
'''
input_str = input()
dimensions=[int(x) for x in input_str.split(',')]
rowNum=dimensions[0]
colNum=dimensions[1]
multilist = [[0 for col in range(colNum)] for row in range(rowNum)]

for row in range(rowNum):
    for col in range(colNum):
        multilist[row][col]= row*col

print(multilist)
#----------------------------------------#

3,5
[[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]


In [ ]:
input_str = input()
dimensions=[int(x) for x in input_str.split(',')]
rowNum=dimensions[0]
colNum=dimensions[1]
multilist = [[0 for col in range(colNum)] for row in range(rowNum)]

3,5


In [ ]:
multilist

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]

In [ ]:
ord('p')

112

#Google Python Crash course

The fractional_part function divides the numerator by the denominator, and returns just the fractional part (a number between 0 and 1). Complete the body of the function so that it returns the right number.
Note: Since division by 0 produces an error, if the denominator is 0, the function should return 0 instead of attempting the division.

In [ ]:
def fractional_part(numerator, denominator):
  if numerator and denominator !=0:
    overall_answer=numerator/denominator
    interger_part=numerator//denominator
    return overall_answer-interger_part
  return 0
	# Operate with numerator and denominator to 
  # keep just the fractional part of the quotient
    


print(fractional_part(5, 5)) # Should be 0
print(fractional_part(5, 4)) # Should be 0.25
print(fractional_part(5, 3)) # Should be 0.66...
print(fractional_part(5, 2)) # Should be 0.5
print(fractional_part(5, 0)) # Should be 0
print(fractional_part(0, 5)) # Should be 0

0.0
0.25
0.6666666666666667
0.5
0
0


In [ ]:
x=0
while x==0:
  x=x/2
  x+=1
  print(x)

1.0


Question 2
Fill in the blanks to make the print_prime_factors function print all the prime factors of a number. A prime factor is a number that is prime and divides another without a remainder.

In [ ]:
def print_prime_factors(number):
  # Start with two, which is the first prime
  factor = 2
  # Keep going until the factor is larger than the number
  while factor <= number:
    # Check if factor is a divisor of number
    if number % factor == 0:
      # If it is, print it and divide the original number
      print(factor)
      number = number / factor
    else:
      # If it's not, increment the factor by one
      factor+=1
      
  return "Done"

print_prime_factors(100)
# Should print 2,2,5,5
# DO NOT DELETE THIS COMMENT

2
2
5
5


'Done'

In [ ]:
def is_power_of_two(n):
  # Check if the number can be divided by two without a remainder
  while n % 2 == 0:
    if n == 0 :
      break;      
    n = n / 2
  # If after dividing by two the number is 1, it's a power of two
  if n == 1:
    return True
  return False
  

print(is_power_of_two(0)) # Should be False
print(is_power_of_two(1)) # Should be True
print(is_power_of_two(8)) # Should be True
print(is_power_of_two(9)) # Should be False

False
True
True
False


4.
Question 4
Fill in the empty function so that it returns the sum of all the divisors of a number, without including it. A divisor is a number that divides into another without a remainder.

In [ ]:
def sum_divisors(n):
  sum = 0
  # Return the sum of all divisors of n, not including n
  num=1
  while num<n:
    if n%num==0:
      sum+=num
    num+=1

  return sum

print(sum_divisors(0))
# 0
print(sum_divisors(3)) # Should sum of 1
# 1
print(sum_divisors(36)) # Should sum of 1+2+3+4+6+9+12+18
# 55
print(sum_divisors(102)) # Should be sum of 2+3+6+17+34+51
# 114


0
1
55
114


The multiplication_table function prints the results of a number passed to it multiplied by 1 through 5. An additional requirement is that the result is not to exceed 25, which is done with the break statement. Fill in the blanks to complete the function to satisfy these conditions.

In [ ]:
def multiplication_table(number):
	# Initialize the starting point of the multiplication table
	multiplier = 1
	# Only want to loop through 5
	while multiplier <= 5:
		result = multiplier*number
		# What is the additional condition to exit out of the loop?
		if result>25 :
			break
		print(str(number) + "x" + str(multiplier) + "=" + str(result))
		# Increment the variable for the loop
		multiplier+= 1

multiplication_table(3) 
# Should print: 3x1=3 3x2=6 3x3=9 3x4=12 3x5=15

multiplication_table(5) 
# Should print: 5x1=5 5x2=10 5x3=15 5x4=20 5x5=25

multiplication_table(8)	
# Should print: 8x1=8 8x2=16 8x3=24

3x1=3
3x2=6
3x3=9
3x4=12
3x5=15
5x1=5
5x2=10
5x3=15
5x4=20
5x5=25
8x1=8
8x2=16
8x3=24


Question 2
Fill in the blanks to make the factorial function return the factorial of n. Then, print the first 10 factorials (from 0 to 9) with the corresponding number. Remember that the factorial of a number is defined as the product of an integer and all integers before it. For example, the factorial of five (5!) is equal to 1*2*3*4*5=120. Also recall that the factorial of zero (0!) is equal to 1.

In [ ]:
def factorial(n):
    result = 1
    for x in range(1,n):
        result = result *x
    return result

for n in range(0,10):
    print(n, factorial(n+1))

0 1
1 1
2 2
3 6
4 24
5 120
6 720
7 5040
8 40320
9 362880


Write a script that prints the first 10 cube numbers (x**3), starting with x=1 and ending with x=10.

In [ ]:
for x in range(1,11):
  print(x**3)

1
8
27
64
125
216
343
512
729
1000


Question 4
Write a script that prints the multiples of 7 between 0 and 100. Print one multiple per line and avoid printing any numbers that aren't multiples of 7. Remember that 0 is also a multiple of 7

In [ ]:
# nums=range(0,101)
for i in range(0,101):
  if i%7==0:
    print(i)

0
7
14
21
28
35
42
49
56
63
70
77
84
91
98


Question 5
The retry function tries to execute an operation that might fail, it retries the operation for a number of attempts.  Currently the code will keep executing the function even if it succeeds. Fill in the blank so the code stops trying after the operation succeeded.

In [ ]:
#@title
# this is the syntax, functions are predefined in their system.
'''def retry(operation, attempts):
  for n in range(attempts):
    if operation():
      print("Attempt " + str(n) + " succeeded")
      break
    else:
      print("Attempt " + str(n) + " failed")

retry(create_user, 3)
retry(stop_service, 5)'''

'def retry(operation, attempts):\n  for n in range(attempts):\n    if operation():\n      print("Attempt " + str(n) + " succeeded")\n      break\n    else:\n      print("Attempt " + str(n) + " failed")\n\nretry(create_user, 3)\nretry(stop_service, 5)'

The function sum_positive_numbers should return the sum of all positive numbers between the number n received and 1. For example, when n is 3 it should return 1+2+3=6, and when n is 5 it should return 1+2+3+4+5=15. Fill in the gaps to make this work:

In [ ]:
def sum_positive_numbers(n):
    # The base case is n being smaller than 1
    if n < 1:      
      
        return 0

    # The recursive case is adding this number to 
    # the sum of the numbers smaller than this one.
    return n + sum_positive_numbers(n-1)

print(sum_positive_numbers(3)) # Should be 6
print(sum_positive_numbers(5)) # Should be 15

6
15


##recursion

In [ ]:
def recursive_function(parameters):
    if base_case_condition(parameters):
        return base_case_value
    recursive_function(modified_parameters)

Question 3
Fill in the blanks to make the is_power_of function return whether the number is a power of the given base. Note: base is assumed to be a positive number. Tip: for functions that return a boolean value, you can return the result of a comparison.

In [ ]:
def is_power_of(number, base):
  # Base case: when number is smaller than base.
  if number < base:
    # If number is equal to 1, it's a power (base**0).
    if number==1:
      return True
    elif number!=1:
      return False
    return number/base

  # Recursive case: keep dividing number by base.
  return is_power_of(number/base,base)

print(is_power_of(8,2)) # Should be True
print(is_power_of(64,4)) # Should be True
print(is_power_of(70,10)) # Should be False

True
True
False


Question 4
The count_users function recursively counts the amount of users that belong to a group in the company system, by going through each of the members of a group and if one of them is a group, recursively calling the function and counting the members. But it has a bug! Can you spot the problem and fix it?

In [ ]:
def count_users(group):
  count = 0
  for member in get_members(group):
    # print('memeber = '+member)
    count += 1
    if is_group(member):
      count += count_users(member)-1
    # else:
      # count+=1
  return count

print(count_users("sales")) # Should be 3
print(count_users("engineering")) # Should be 8
print(count_users("everyone")) # Should be 18

Question 5
Implement the sum_positive_numbers function, as a recursive function that returns the sum of all positive numbers between the number n received and 1. For example, when n is 3 it should return 1+2+3=6, and when n is 5 it should return 1+2+3+4+5=15.

In [ ]:
def sum_positive_numbers(n):
  return 0

print(sum_positive_numbers(3)) # Should be 6
print(sum_positive_numbers(5)) # Should be 15

#Module 3 Graded Assessment

Fill in the blanks of this code to print out the numbers 1 through 7.

In [ ]:
number = 1
while number <= 7:
	print(number, end=" ")
	number+=1

1 2 3 4 5 6 7 

2.
Question 2
The show_letters function should print out each letter of a word on a separate line. Fill in the blanks to make that happen.

In [ ]:
def show_letters(word):
	for x in word:
		print(x)

show_letters("Hello")
# Should print one line per letter

H
e
l
l
o


3.
Question 3
Complete the function digits(n) that returns how many digits the number has. For example: 25 has 2 digits and 144 has 3 digits. Tip: you can figure out the digits of a number by dividing it by 10 once per digit until there are no digits left.

In [ ]:
def digits(n):
    count = 0
    if n == 0:
        count=1
    while (n>0):
        count += 1
        n//=10
    return count

print(digits(25))   # Should print 2
print(digits(144))  # Should print 3
print(digits(1000)) # Should print 4
print(digits(0))    # Should print 1

2
3
4
1


Question 4
This function prints out a multiplication table (where each number is the result of multiplying the first number of its row by the number at the top of its column). Fill in the blanks so that calling multiplication_table(1, 3) will print out:

1 2 3 

2 4 6 

3 6 9

In [ ]:
def multiplication_table(start, stop):
	for x in range(1,start+1):  [1,2,
		for y in range(1,stop+1):  [1,2,3,]
			print(str(x*y), end=" ")
		print()

# multiplication_table(1, 3)
multiplication_table(2, 4)
# Should print the multiplication table shown above

1 2 3 4 
2 4 6 8 


In [ ]:
a=range(1,2)
a

range(1, 2)

In [ ]:
for i in range(1,3):
  print(i)

1
2


In [ ]:
def multiplication_table(start, stop):
	for x in range(start,stop+1): 
		for y in range(start,stop+1):  
			print(str(x*y), end=" ")
		print()

multiplication_table(1, 3)
# Should print the multiplication table shown above

1 2 3 
2 4 6 
3 6 9 


5.
Question 5
The counter function counts down from start to stop when start is bigger than stop, and counts up from start to stop otherwise. Fill in the blanks to make this work correctly.

In [ ]:
def counter(start, stop):
	x = start
	if x>stop:
		return_string = "Counting down: "
		while x >= stop:
			return_string += str(x)
			if x!=stop:
				return_string += ","
			x-=1
	else:
		return_string = "Counting up: "
		while x <= stop:
			return_string += str(x)
			if x!=stop:
				return_string += ","
			x+=1
	return return_string

print(counter(1, 10)) # Should be "Counting up: 1,2,3,4,5,6,7,8,9,10"
print(counter(2, 1)) # Should be "Counting down: 2,1"
print(counter(5, 5)) # Should be "Counting up: 5"

Counting up: 1,2,3,4,5,6,7,8,9,10
Counting down: 2,1
Counting up: 5


The even_numbers function returns a space-separated string of all positive numbers that are divisible by 2, up to and including the maximum that's passed into the function. For example, even_numbers(6) returns “2 4 6”. Fill in the blank to make this work.

In [ ]:
def even_numbers(maximum):
  return_string = ""
  for x in range(1,maximum+1):
    if x%2==0:
      return_string += str(x) + " "
  return return_string.strip()

print(even_numbers(6))  # Should be 2 4 6
print(even_numbers(10)) # Should be 2 4 6 8 10
print(even_numbers(1))  # No numbers displayed
print(even_numbers(3))  # Should be 2
print(even_numbers(0))  # No numbers displayed

2 4 6
2 4 6 8 10

2



In [ ]:
for x in range(1, 10, 3):
    print(x)

1
4
7


In [ ]:
for x in range(10):
    for y in range(x):
        print(y)

0
0
1
0
1
2
0
1
2
3
0
1
2
3
4
0
1
2
3
4
5
0
1
2
3
4
5
6
0
1
2
3
4
5
6
7
0
1
2
3
4
5
6
7
8


In [ ]:
def votes(params):
	for vote in params:
	    print("Possible option:" + vote)


In [ ]:
votes(['yes', 'no', 'maybe'])

Possible option:yes
Possible option:no
Possible option:maybe


Want to give it a go yourself? Be my guest! Modify the first_and_last function so that it returns True if the first letter of the string is the same as the last letter of the string, False if they’re different. Remember that you can access characters using message[0] or message[-1].
Be careful how you handle the empty string, which should return True since nothing is equal to nothing.

In [ ]:
def first_and_last(message):
    if message=="":
        return True
    elif message[0]==message[-1]:
        return True
    else:
        return False

print(first_and_last("else"))
print(first_and_last("tree"))
print(first_and_last(""))

True
False
True


Practice Quiz: Strings

The is_palindrome function checks if a string is a palindrome. A palindrome is a string that can be equally read from left to right or right to left, omitting blank spaces, and ignoring capitalization. Examples of palindromes are words like kayak and radar, and phrases like "Never Odd or Even". Fill in the blanks in this function to return True if the passed string is a palindrome, False if not.

In [ ]:
def is_palindrome(input_string):
	# We'll create two strings, to compare them
	new_string = ""
	reverse_string = ""
	# Traverse through each letter of the input string
	for ___:
		# Add any non-blank letters to the 
		# end of one string, and to the front
		# of the other string. 
		if ___:
			new_string = ___
			reverse_string = ___
	# Compare the strings
	if ___:
		return True
	return False

print(is_palindrome("Never Odd or Even")) # Should be True
print(is_palindrome("abc")) # Should be False
print(is_palindrome("kayak")) # Should be True

In [1]:
astring='My name is Sajid'

In [2]:
for i in astring:
  print(i)

M
y
 
n
a
m
e
 
i
s
 
S
a
j
i
d


In [3]:
for i in astring.strip():
  print(i)

M
y
 
n
a
m
e
 
i
s
 
S
a
j
i
d


In [7]:
bstring=astring.lstrip()

In [8]:
bstring

'My name is Sajid'

In [12]:
''.join(astring)

'My name is Sajid'

In [17]:
slist=astring.split(' ')
slist

['My', 'name', 'is', 'Sajid']

In [18]:
for i in range(len(slist)):
  print(i)

0
1
2
3


In [19]:
def is_palindrome(input_string):
    astring=''
    bstring=''
    for i in input_string:
        if i!=' ':
            astring+=i.lower()
    for i in range(1,len(astring)+1):
        bstring+=astring[-i]
    if astring==bstring:
        return True
    else:
        return False

print(is_palindrome("Never Odd or Even")) # Should be True
print(is_palindrome("abc")) # Should be False
print(is_palindrome("kayak")) # Should be True

True
False
True


Question 4
Fill in the gaps in the nametag function so that it uses the format method to return first_name and the first initial of last_name followed by a period. For example, nametag("Jane", "Smith") should return "Jane S."

In [20]:
def nametag(first_name, last_name):
	return("{} {:.1s}.".format(first_name,last_name))

print(nametag("Jane", "Smith")) 
# Should display "Jane S." 
print(nametag("Francesco", "Rinaldi")) 
# Should display "Francesco R." 
print(nametag("Jean-Luc", "Grand-Pierre")) 
# Should display "Jean-Luc G." 

Jane S.
Francesco R.
Jean-Luc G.


The replace_ending function replaces the old string in a sentence with the new string, but only if the sentence ends with the old string. If there is more than one occurrence of the old string in the sentence, only the one at the end is replaced, not all of them. For example, replace_ending("abcabc", "abc", "xyz") should return abcxyz, not xyzxyz or xyzabc. The string comparison is case-sensitive, so replace_ending("abcabc", "ABC", "xyz") should return abcabc (no changes made). 

In [42]:
def replace_ending(sentence, old, new):
	# Check if the old string is at the end of the sentence 
	if old in sentence:
		# Using i as the slicing index, combine the part
		# of the sentence up to the matched string at the 
		# end with the new string
		i = sentence.index(old)
		new_sentence = sentence[:i]+sentence[i-1]+new
		return new_sentence

	# Return the original sentence if there is no match 
	return sentence
	
print(replace_ending("It's raining cats and cats", "cats", "dogs")) 
# Should display "It's raining cats and dogs"
print(replace_ending("She sells seashells by the seashore", "seashells", "donuts")) 
# Should display "She sells seashells by the seashore"
print(replace_ending("The weather is nice in May", "may", "april")) 
# Should display "The weather is nice in May"
print(replace_ending("The weather is nice in May", "May", "April")) 
# Should display "The weather is nice in April"





It's raining  dogs
She sells  donuts
The weather is nice in May
The weather is nice in  April


In [ ]:
def replace_ending(sentence, old, new):
	# Check if the old string is at the end of the sentence 
	if old in sentence:
    
		# Using i as the slicing index, combine the part
		# of the sentence up to the matched string at the 
		# end with the new string
		i = ___
		new_sentence = ___
		return new_sentence

	# Return the original sentence if there is no match 
	return sentence

In [27]:
old='abc@gmail.com'
new='hotmail.com'

In [28]:
'@' in old

True

In [29]:
old.index('@')

3

In [31]:
old[:3]+"@"+new

'abc@hotmail.com'

In [33]:
def new_email(email,old,new):
  if old in email:
    index=email.index(old)
    return email[:index]+new
  else:
    return email

In [34]:
new_email('abc@gmail.com','@gmail.com','@hotmail.com')

'abc@hotmail.com'

In [36]:
new_email('abcd@gmail.com','@gmail.com','@hotmail.com')

'abcd@hotmail.com'

In [58]:
def replace_ending(sentence, old, new):
  # Check if the old string is at the end of the sentence 
  if sentence.endswith(old):
    # Using i as the slicing index, combine the part
    # of the sentence up to the matched string at the 
    # end with the new string
    i = sentence.rfind(old)
    new_sentence = sentence[:i]+new
    return new_sentence
    # Return the original sentence if there is no match 
  return sentence
print(replace_ending("It's raining cats and cats", "cats", "dogs")) 
# Should display "It's raining cats and dogs"
print(replace_ending("She sells seashells by the seashore", "seashells", "donuts")) 
# Should display "She sells seashells by the seashore"
print(replace_ending("The weather is nice in May", "may", "april")) 
# Should display "The weather is nice in May"
print(replace_ending("The weather is nice in May", "May", "April")) 
# Should display "The weather is nice in April"





It's raining cats and dogs
She sells seashells by the seashore
The weather is nice in May
The weather is nice in April


In [46]:
aa="It's raining cats and cats"

In [47]:
aa.index('cats')

13

In [48]:
aa.count('cats')

2